stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -396,6 +396,85 @@ function _rdiv!(Dc::Diagonal, Db::Diagonal, Da::Diagonal)
 end
 ldiv!(Dc::Diagonal, Da::Diagonal, Db::Diagonal) = Diagonal(ldiv!(Dc.diag, Da, Db.diag))
 
+# optimizations for (Sym)Tridiagonal and Diagonal
+function (\)(D::Diagonal, S::SymTridiagonal)
+    T = typeof(oneunit(eltype(D)) \ oneunit(eltype(S)))
+    # this is essentially copy_oftype, but we need to handle the case when S.ev has the
+    # same length as S.dv
+    dl = copyto!(similar(S.ev, T, length(S.dv)-1), _evview(S))
+    d  = copy_oftype(S.dv, T)
+    du = copy(dl)
+    ldiv!(D, Tridiagonal(dl, d, du))
+end
+function (\)(D::Diagonal, T::Tridiagonal)
+    Td = typeof(oneunit(eltype(D)) \ oneunit(eltype(T)))
+    ldiv!(D, copy_oftype(T, Td))
+end
+function ldiv!(Tr::Tridiagonal, D::Diagonal, T::Tridiagonal)
+    m = length(T.d)
+    dd = D.diag
+    if (k = length(dd)) != m
+        throw(DimensionMismatch("diagonal matrix is $k by $k but right hand side has $m rows"))
+    end
+    if length(Tr.d) != m
+        throw(DimensionMismatch("target matrix size $(size(Tr)) does not match input matrix size $(size(T))"))
+    end
+    j = findfirst(iszero, dd)
+    isnothing(j) || throw(SingularException(j))
+    ddj = dd[1]
+    Tr.d[1] = ddj \ T.d[1]
+    Tr.du[1] = ddj \ T.du[1]
+    for j in 2:m-1
+        ddj = dd[j]
+        Tr.dl[j-1] = ddj \ T.dl[j-1]
+        Tr.d[j]  = ddj \ T.d[j]
+        Tr.du[j] = ddj \ T.du[j]
+    end
+    ddj = dd[m]
+    Tr.dl[m-1] = ddj \ T.dl[m-1]
+    Tr.d[m] = ddj \ T.d[m]
+    return Tr
+end
+
+function (/)(S::SymTridiagonal, D::Diagonal)
+    T = typeof(oneunit(eltype(S)) / oneunit(eltype(D)))
+    # this is essentially copy_oftype, but we need to handle the case when S.ev has the
+    # same length as S.dv
+    dl = copyto!(similar(S.ev, T, length(S.dv)-1), _evview(S))
+    d  = copy_oftype(S.dv, T)
+    du = copy(dl)
+    rdiv!(Tridiagonal(dl, d, du), D)
+end
+function (/)(T::Tridiagonal, D::Diagonal)
+    Td = typeof(oneunit(eltype(T)) / oneunit(eltype(D)))
+    rdiv!(copy_oftype(T, Td), D)
+end
+function _rdiv!(Tr::Tridiagonal, T::Tridiagonal, D::Diagonal)
```

Comment:
In the current state, this gets called in the 2-arg `rdiv!`, which forwards to `_rdiv!(similar(T), T, D)`. But I'm working on including the last comments, and in that version I call 3-arg `_rdiv!` directly to avoid copying `T`.